90

# Question 1: Decision Trees

In [1]:
import pandas as pd
import numpy as np
df = pd.read_csv("../Crime Prediction Data/communities-crime-clean.csv")

In [2]:
df[:10]

state      communityname  fold  population  householdsize  racepctblack  \
0      1      Alabastercity     7        0.01           0.61          0.21   
1      1  AlexanderCitycity    10        0.01           0.41          0.55   
2      1       Annistoncity     3        0.03           0.34          0.86   
3      1         Athenscity     8        0.01           0.38          0.35   
4      1         Auburncity     1        0.04           0.37          0.32   
5      1       Bessemercity     6        0.04           0.44          1.00   
6      1     Birminghamcity     2        0.41           0.37          1.00   
7      1        Cullmancity     1        0.01           0.30          0.00   
8      1         Daphnecity     7        0.00           0.39          0.31   
9      1        Decaturcity    10        0.06           0.39          0.32   

   racePctWhite  racePctAsian  racePctHisp  agePct12t21         ...           \
0          0.83          0.02         0.01         0.41         ...            
1          0.57          0.01         0.00         0.47         ...            
2          0.30          0.04         0.01         0.41         ...            
3          0.71          0.04         0.01         0.39         ...            
4          0.70          0.21         0.02         1.00         ...            
5          0.10          0.00         0.00         0.43         ...            
6          0.02          0.03         0.01         0.41         ...            
7          0.99          0.02         0.01         0.38         ...            
8          0.75          0.02         0.02         0.35         ...            
9          0.73          0.04         0.01         0.39         ...            

   PctForeignBorn  PctBornSameState  PctSameHouse85  PctSameCity85  \
0            0.03              0.70            0.40           0.34   
1            0.00              0.93            0.66           0.82   
2            0.04              0.77            0.59           0.70   
3            0.03              0.78            0.56           0.67   
4            0.12              0.49            0.12           0.00   
5            0.00              0.96            0.74           0.95   
6            0.03              0.90            0.64           0.86   
7            0.00              0.82            0.54           0.72   
8            0.03              0.60            0.46           0.40   
9            0.03              0.71            0.47           0.61   

   PctSameState85  LandArea  PopDens  PctUsePubTrans  LemasPctOfficDrugUn  \
0            0.57      0.05     0.06            0.01                 0.00   
1            0.84      0.11     0.03            0.01                 0.00   
2            0.64      0.06     0.11            0.04                 0.00   
3            0.71      0.09     0.05            0.00                 0.00   
4            0.15      0.09     0.09            0.01                 0.00   
5            0.89      0.11     0.07            0.13                 0.00   
6            0.82      0.43     0.15            0.20                 0.38   
7            0.76      0.04     0.07            0.01                 0.00   
8            0.54      0.03     0.08            0.01                 0.00   
9            0.60      0.14     0.09            0.01                 0.00   

   ViolentCrimesPerPop  
0                 0.06  
1                 0.14  
2                 1.00  
3                 0.23  
4                 0.15  
5                 1.00  
6                 1.00  
7                 0.16  
8                 0.05  
9                 0.22  

[10 rows x 104 columns]

In [3]:
df['highCrime'] = (df['ViolentCrimesPerPop'] > 0.1)

In [4]:
df.highCrime.value_counts()

True     1250
False     743
Name: highCrime, dtype: int64

In [5]:
#Checking if the above counts are right
p=sum(df['ViolentCrimesPerPop'][df['ViolentCrimesPerPop']>0.1].value_counts())

In [6]:
p # right

1250

In [7]:
#percentage of positive and negative:
p1= df.highCrime.value_counts()

In [8]:
p1

True     1250
False     743
Name: highCrime, dtype: int64

In [9]:
total=p1[0]+p1[1]

In [10]:
#Positive percentage:
percentage_positive = p1[1]/total*100
percentage_positive  
# Output: 62.719518314099346

62.719518314099346

In [11]:
#Negative percentage:
percentage_negative = p1[0]/total*100
percentage_negative  
# Output: 37.280481685900654

37.280481685900654

Percentage of positive instances : 62.72%
Percentage of negative instances : 37.28%

In [12]:
#Dividing data into train and test set
X = df.drop(['communityname','ViolentCrimesPerPop','highCrime'],axis=1)
y = df['highCrime']
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split( X, y, test_size=0.2, random_state=42)


>There are some other columns that cannot be considered as features, such as fold, state.

>Why did you split the data into train and test split? This was not asked.

Predicting on X_test

In [13]:
#Building Decision tree classifier
from sklearn import tree
dtc = tree.DecisionTreeClassifier(criterion = 'entropy')
dtc = dtc.fit(X_train, y_train)

y=dtc.predict(X_test)

In [14]:
from sklearn.metrics import accuracy_score,precision_score,recall_score, f1_score
print("Accuracy= ",accuracy_score(y_test,y))
print("Precision= ",precision_score(y_test,y)) 
print("Recall= ",recall_score(y_test,y)) 
print("f1-score= ",f1_score(y_test,y))


Accuracy=  0.779448621554
Precision=  0.848979591837
Recall=  0.803088803089
f1-score=  0.825396825397


predicting on X_train

In [15]:
y1=dtc.predict(X_train)

In [16]:
from sklearn.metrics import accuracy_score,precision_score,recall_score, f1_score
print("Accuracy= ",accuracy_score(y_train,y1))
print("Precision= ",precision_score(y_train,y1)) 
print("Recall= ",recall_score(y_train,y1)) 
print("f1-score= ",f1_score(y_train,y1))

Accuracy=  1.0
Precision=  1.0
Recall=  1.0
f1-score=  1.0


In [17]:
dimensions= list(X)


In [19]:
dtc.feature_importances_

array([ 0.01517591,  0.01043863,  0.01025166,  0.00797971,  0.00525841,
        0.08782887,  0.        ,  0.06699369,  0.        ,  0.00522777,
        0.00932864,  0.02020872,  0.        ,  0.        ,  0.00493345,
        0.00665216,  0.00442474,  0.        ,  0.00654056,  0.01271924,
        0.00255716,  0.00370645,  0.        ,  0.00815736,  0.01114912,
        0.        ,  0.        ,  0.01418764,  0.00461075,  0.00384795,
        0.        ,  0.01018966,  0.01536211,  0.00478531,  0.00376529,
        0.        ,  0.00593365,  0.01342649,  0.        ,  0.00904565,
        0.00757717,  0.        ,  0.01550128,  0.01447228,  0.00672013,
        0.00425537,  0.31632095,  0.00555131,  0.00705875,  0.00743447,
        0.        ,  0.00856699,  0.        ,  0.0070006 ,  0.01437725,
        0.00791747,  0.00994184,  0.        ,  0.00461081,  0.        ,
        0.00296969,  0.        ,  0.01084518,  0.0085545 ,  0.        ,
        0.00131132,  0.        ,  0.00421877,  0.00529336,  0.00

In [21]:
significance = dtc.feature_importances_
f=zip(significance, dimensions)
l=sorted(f, reverse=True)
l[:5]

[(0.31632095236026891, 'PctKids2Par'),
 (0.087828869739432186, 'racePctWhite'),
 (0.066993685521093083, 'racePctHisp'),
 (0.020208723910586056, 'agePct65up'),
 (0.017493834030438686, 'HousVacant')]

PctKids2Par, racePctWhite, racePctHisp, HousVacant, blackPerCap are the most significant variables in this model. 
PctKids2Par- percentage of kids in family housing with two parents (numeric - decimal)
racePctWhite- percentage of population that is caucasian (numeric - decimal)
racePctHisp- percentage of population that is of hispanic heritage
HousVacant- number of vacant households (numeric - decimal)
blackPerCap- per capita income for african americans (numeric - decimal)

Out of the above 5 factors, the first (PctKids2Par), makes sense. The higher percentage of kids in family housing could mean 
higher targets for crime and eventually higher crime rate.



In [22]:
from sklearn.model_selection import cross_val_score
accuracy =  cross_val_score(dtc, X_train, y_train, cv=10)
precision =  cross_val_score(dtc, X_train, y_train, cv=10 , scoring = 'precision')
recall =  cross_val_score(dtc, X_train, y_train, cv=10 , scoring = 'recall')
print ("Mean Accuracy = ", accuracy.mean())
print ("Mean Recall = ", recall.mean())
print ("Mean Precision = ", precision.mean())


Mean Accuracy =  0.76101598207
Mean Recall =  0.793141414141
Mean Precision =  0.821633920357


The result for accuracy after cross-validation is 75.59% as compared to 100% obtained on X_train. 
The reason for getting 100% earlier was that the model was trained on the same data set as it was tested on. 10-fold cross validation generates "out of sample" data, and when the model is tested on it, the average accuracy obviously redcues due to new and varying sample instances. 

> Now the problem is, you could use the entire data for X-validation. But you used only 4/5 of it. Generally, we don't spare data, when we do X-validation, because X-validation internally spare part of the data as test and does not use it in the training at every iteration.

# Question 2: Linear Classification 

In [23]:
from sklearn.naive_bayes import GaussianNB
gnb=GaussianNB()
gnb.fit(X_train, y_train)
y=gnb.predict(X_test)

from sklearn.metrics import accuracy_score,precision_score,recall_score, f1_score
print("Accuracy= ",accuracy_score(y_test,y))
print("Precision= ",precision_score(y_test,y)) 
print("Recall= ",recall_score(y_test,y)) 
print("f1-score= ",f1_score(y_test,y))


#Accuracy=  0.759398496241
#Precision=  0.922279792746
#Recall=  0.687258687259
#f1-score=  0.787610619469

Accuracy=  0.759398496241
Precision=  0.922279792746
Recall=  0.687258687259
f1-score=  0.787610619469


In [56]:
from sklearn.model_selection import cross_val_score
accuracy =  cross_val_score(gnb, X_train, y_train, cv=10)
precision =  cross_val_score(gnb, X_train, y_train, cv=10 , scoring = 'precision')
recall =  cross_val_score(gnb, X_train, y_train, cv=10 , scoring = 'recall')
print ("Mean Accuracy = ", accuracy.mean())
print ("Mean Recall = ", recall.mean())
print ("Mean Precision = ", precision.mean())




#Mean Accuracy =  0.77728204715
#Mean Recall =  0.695262626263
#Mean Precision =  0.929646088276




Mean Accuracy =  0.77728204715
Mean Recall =  0.695262626263
Mean Precision =  0.929646088276


In [114]:
l= X.columns
d = pd.DataFrame({"Dimension": [], "Significance": []})

for i in l:
    means = df.groupby('highCrime')[i].mean()
    variance = df.groupby('highCrime')[i].var()
    
    d = d.append({"Dimension": i, "Significance": abs(means[0] - means[1])/ sum(np.sqrt(variance))}, ignore_index=True)
 
d.iloc[np.argsort(d['Significance'])[::-1][0:5], ]



#Most significant variables identified = PctKids2Par,PctFam2Par, racePctWhite, PctIlleg, FemalePctDiv

Dimension  Significance
46   PctKids2Par      0.809336
45    PctFam2Par      0.745162
5   racePctWhite      0.734884
52      PctIlleg      0.708929
42  FemalePctDiv      0.693604

PctKids2Par, PctFam2Par, racePctWhite, PctIlleg,FemalePctDiv are the most significant variables identified
PctKids2Par- percentage of kids in family housing with two parents (numeric - decimal)
PctFam2Par - percentage of families (with kids) that are headed by two parents (numeric - decimal)
racePctWhite - percentage of population that is caucasian (numeric - decimal)
PctIlleg - percentage of kids born to never married (numeric - decimal) 
FemalePctDiv - percentage of females who are divorced (numeric - decimal)

Out of the above 6 factors, the first (PctKids2Par) and the 2nd (PctFam2Par) make sense and is consistent with what was observed in the earlier decision tree. The higher percentage of kids in family housing with 2 parents could mean 
higher targets for crime and eventually higher crime rate. And higher the percentage of such households would linearly increase/decrease the targets.



> It would be nice to mention the formula, and discuss why it make sense to calculate the distance between two means and then to divide by the summation of standard deviations.

> Just a presentation issue: Some subsection titles would be helpful.

In [24]:
from sklearn.svm import LinearSVC
svc =LinearSVC()
svc.fit(X_train, y_train)

LinearSVC(C=1.0, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0)

In [25]:
from sklearn.model_selection import cross_val_score
accuracy = cross_val_score(svc, X_train, y_train,cv=10,scoring="accuracy")
recall = cross_val_score(svc,X_train, y_train,cv=10,scoring="recall")
precision = cross_val_score(svc, X_train, y_train,cv=10,scoring="precision")
print ("Mean Accuracy = ", accuracy.mean())
print ("Mean Recall = ", recall.mean())
print ("Mean Precision = ", precision.mean())


#Mean Accuracy =  0.796103094847
#Mean Recall =  0.821282828283
#Mean Precision =  0.84267155864

Mean Accuracy =  0.781777974726
Mean Recall =  0.87397979798
Mean Precision =  0.88491157796


In [27]:
dimensions =list(X_train)
weights =svc.coef_
weights=weights[0]
f= zip(weights, dimensions)
l=sorted(f, reverse=True)
l[:5]

[(1.0715027620433351, 'MalePctDivorce'),
 (0.94507439573333485, 'racepctblack'),
 (0.92514278499990166, 'PctOccupMgmtProf'),
 (0.91124938062367722, 'agePct12t21'),
 (0.90133011980324307, 'racePctHisp')]

MalePctDivorce, racepctblack, PctOccupMgmtProf, agePct12t21, racePctHisp are the most significant variables identified.
MalePctDivorce - percentage of population who are divorced (numeric - decimal)
racepctblack - percentage of population that is african american (numeric - decimal)
PctOccupMgmtProf - percentage of people 16 and over who are employed in management or professional occupations (numeric - decimal)
agePct12t21 - percentage of population that is 12-21 in age (numeric - decimal)
racePctHisp- percentage of population that is of hispanic heritage (numeric - decimal)

Out of the above features identified, the ones that make sense are PctOccupMgmtProf and agePct12t21. This is because, people in this 
age range and occupations are likely to be outside their homes during the day, hence they could be potential targets. Difference 
in the values of these features will make an impact on the predcited variable.


No parameters between this model and Gaussian Naive Bayes models are similar. Only one paramter, which is racePctHisp,
is common between this model and decision trees. 

> This is an important observation and it would be nice to elaborate why SVM and GNB make use of different attributes.

# Question 3: Regression

In [76]:
X = df.drop(['communityname','ViolentCrimesPerPop','highCrime'],axis=1)
y = df['ViolentCrimesPerPop']
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split( X, y, test_size=0.2, random_state=42)

> Same two remarks as above: There are some other columns to be discarded and why did you arbitrarily split the data into training and test sets?

In [28]:
from sklearn.linear_model import LinearRegression
lm=LinearRegression()
lm.fit(X_train,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [29]:
from sklearn.metrics import mean_squared_error
mse =  cross_val_score(lm, X_train, y_train, cv=10 , scoring = 'neg_mean_squared_error')  
print ("Mean of MSE: ", abs(mse.mean()))

Mean of MSE:  0.133018201262


> There is a mistake here: You train lm on the training set (X_train, y_train). Then you call cross_val_score() routine to run X-validation on your model. This routine internally trains the input model, which you give as your previously trained model lm. I am not sure. This may or may not be a problem. But if linear regression model here allows incremental training, that is, fit() function, once called for the second time, trains the model where it was left off at the end of the first training, then this X-validation you called here is not independent from your initial training.

In [33]:
y =lm.predict(X_train)
m=sum((y-y_train)**2)/len(y_test)
print("MSE on the training set:  ",m)

MSE on the training set:   0.4601785489173281


> I think this result deserves some discussion. The X-validation MSE is better than the individual MSE. Isn't it surprising? X-validation uses subsets of the same data. So the training set shrank but the MSE has increased, especially being test on a data that was not used in training. This difference is quite interesting.

In [34]:
coeffecients = lm.coef_
s = coeffecients >= 0
abs_coeffecients=abs(lm.coef_)
i = np.argsort(abs_coeffecients)[::-1]
for f in range(10):
    print("%s- %f - %s" % (X_train.columns[i[f]], abs_coeffecients[i[f]], s[i[f]]))

population- 3.706893 - True
numbUrban- 2.378423 - False
PctKids2Par- 1.459143 - False
PctRecImmig10- 1.114016 - False
pctWInvInc- 1.007788 - False
PctOccupMgmtProf- 0.930633 - True
MedRent- 0.885291 - False
PersPerOccupHous- 0.847443 - True
NumIlleg- 0.824601 - False
PctRecImmig5- 0.811896 - True


Predcitors of High crimes:

1. PctHousOwnOcc
2. MalePctDivorce
3. PctRecImmig8
4. PersPerOccupHous
5. MedRent


Predictors of low crimes:

1. PctPersOwnOccup
2. TotalPctDiv
3. whitePerCap
4. OwnOccLowQuart
5. numbUrban

In [35]:
from sklearn.linear_model import RidgeCV
rm = RidgeCV(alphas=(10,1,0.1,0.01,0.001),cv=10)
rm.fit(X_train,y_train)

RidgeCV(alphas=(10, 1, 0.1, 0.01, 0.001), cv=10, fit_intercept=True,
    gcv_mode=None, normalize=False, scoring=None, store_cv_values=False)

In [36]:
m =  cross_val_score(rm, X_train, y_train, cv=10 , scoring = 'neg_mean_squared_error')  
print ("Mean of MSE: ", abs(m.mean()))

Mean of MSE:  0.129886377108


> Same problem I described above.

In [37]:
y =rm.predict(X_train)
m=sum((y-y_train)**2)/len(y_test)
print("MSE on the training set:  ",m)

MSE on the training set:   0.46879732884269815


In [86]:
print("Best Alpha = ",rm.alpha_)

Best Alpha =  1


MSE for Linear regression: 0.0192
MSE for Ridge regression: 0.0189
The mean squared error can be interpreted as the closeness of the data points with the predcited regression line. This explains that Ridge regression is able to redcue overfitting to some extent compared to the result of linear regression.

> Two things: (i) This is an extremely small difference (.0003) for a value that can range from 0 to 1. So it is hard to say the ridge regression helped to reduce overfitting. (ii) I ran your code and the values that I got are quite different than what you indicate here. How did you get these values?

In [38]:
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree=2)
x_p =poly_reg.fit_transform(X_train)

In [39]:
plm=LinearRegression()
qlr_cv=cross_val_score(plm, x_p, y_train,cv=10,scoring="neg_mean_squared_error")
print ("Mean of MSE: ", abs(m.mean()))

AttributeError: 'float' object has no attribute 'mean'

> I ran this cell and got an exception. I copy this cell below, edit it and run again.

In [40]:
plm=LinearRegression()
qlr_cv=cross_val_score(plm, x_p, y_train,cv=10,scoring="neg_mean_squared_error")
print ("Mean of MSE: ", abs(qlr_cv.mean()))

Mean of MSE:  0.802500478325


> Here we get the result.

In [41]:
plm.fit(x_p,y_train)
y=plm.predict(x_p)
m=sum((y-y_train)**2)/len(y_train)
print("MSE on the training set = ",m)

MSE on the training set =  1.7711927142756484e-24


MSE of quadratic model after 10 cross validation: 0.0676

MSE of linear model after 10 cross validation: 0.0189

Since the MSE of quadratic model is higher than the linear model, we conclude that the linear model is better than the quadratic model 

> I checked results in your original submission. These are the values. But I have no idea how you got them. Did you edit the data file by any chance?

# Question 4: Dirty data

In [42]:
df_full = pd.read_csv('../Crime Prediction Data/communities-crime-full.csv')
df_full[:10]

state county community        communityname  fold  population  \
0      8      ?         ?         Lakewoodcity     1        0.19   
1     53      ?         ?          Tukwilacity     1        0.00   
2     24      ?         ?         Aberdeentown     1        0.00   
3     34      5     81440  Willingborotownship     1        0.04   
4     42     95      6096    Bethlehemtownship     1        0.01   
5      6      ?         ?    SouthPasadenacity     1        0.02   
6     44      7     41500          Lincolntown     1        0.01   
7      6      ?         ?            Selmacity     1        0.01   
8     21      ?         ?        Hendersoncity     1        0.03   
9     29      ?         ?          Claytoncity     1        0.01   

   householdsize  racepctblack  racePctWhite  racePctAsian  \
0           0.33          0.02          0.90          0.12   
1           0.16          0.12          0.74          0.45   
2           0.42          0.49          0.56          0.17   
3           0.77          1.00          0.08          0.12   
4           0.55          0.02          0.95          0.09   
5           0.28          0.06          0.54          1.00   
6           0.39          0.00          0.98          0.06   
7           0.74          0.03          0.46          0.20   
8           0.34          0.20          0.84          0.02   
9           0.40          0.06          0.87          0.30   

          ...           LandArea  PopDens  PctUsePubTrans  PolicCars  \
0         ...               0.12     0.26            0.20       0.06   
1         ...               0.02     0.12            0.45          ?   
2         ...               0.01     0.21            0.02          ?   
3         ...               0.02     0.39            0.28          ?   
4         ...               0.04     0.09            0.02          ?   
5         ...               0.01     0.58            0.10          ?   
6         ...               0.05     0.08            0.06          ?   
7         ...               0.01     0.33            0.00          ?   
8         ...               0.04     0.17            0.04          ?   
9         ...               0.00     0.47            0.11          ?   

   PolicOperBudg  LemasPctPolicOnPatr  LemasGangUnitDeploy  \
0           0.04                  0.9                  0.5   
1              ?                    ?                    ?   
2              ?                    ?                    ?   
3              ?                    ?                    ?   
4              ?                    ?                    ?   
5              ?                    ?                    ?   
6              ?                    ?                    ?   
7              ?                    ?                    ?   
8              ?                    ?                    ?   
9              ?                    ?                    ?   

   LemasPctOfficDrugUn  PolicBudgPerPop  ViolentCrimesPerPop  
0                 0.32             0.14                 0.20  
1                 0.00                ?                 0.67  
2                 0.00                ?                 0.43  
3                 0.00                ?                 0.12  
4                 0.00                ?                 0.03  
5                 0.00                ?                 0.14  
6                 0.00                ?                 0.03  
7                 0.00                ?                 0.55  
8                 0.00                ?                 0.53  
9                 0.00                ?                 0.15  

[10 rows x 128 columns]

In [43]:
df_full['highCrime'] = (df_full['ViolentCrimesPerPop'] > 0.1)
df_full.highCrime.value_counts()

True     1251
False     743
Name: highCrime, dtype: int64

Missing value treatment1 : replacing missing values with mean of the respective feature

In [44]:
X = df_full.drop(['communityname','ViolentCrimesPerPop','highCrime'],axis=1)
X=X.apply(pd.to_numeric,errors='coerce')
X = X.fillna(X.mean()).astype(int).astype(int)
y = df_full['highCrime']

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split( X, y, test_size=0.2, random_state=42)

> This cell wasn't run. Why not?

In [45]:
#Building Decision tree classifier
from sklearn import tree
dtc = tree.DecisionTreeClassifier(criterion = 'entropy')
dtc = dtc.fit(X_train, y_train)

y=dtc.predict(X_test)

In [46]:
from sklearn.model_selection import cross_val_score
accuracy =  cross_val_score(dtc, X_train, y_train, cv=10)
precision =  cross_val_score(dtc, X_train, y_train, cv=10 , scoring = 'precision')
recall =  cross_val_score(dtc, X_train, y_train, cv=10 , scoring = 'recall')
print ("Mean Accuracy = ", accuracy.mean())
print ("Mean Recall = ", recall.mean())
print ("Mean Precision = ", precision.mean())

Mean Accuracy =  0.757844105238
Mean Recall =  0.789090909091
Mean Precision =  0.819001472944


>Mean Accuracy =  0.76101598207 
>Mean Recall =  0.793141414141
>Mean Precision =  0.821633920357
>These results are very close, but not identical. They are not identical, because there is a random seed in the cross_val_score routine and we cannot set it. This routing calls StratifiedKFold which requires the random seed (See cross_val_score doc).

In [47]:
X_train.shape

(1594, 102)

In [48]:
X.shape

(1994, 126)

> Your new X has 126 columns, whereas your previous X_train has 102. You loaded the dirty data, and you did not use it. Instead you used the clean data. Why?

> Let's see what these results would be if you used dirty data:

In [49]:
'''Caner added this cell, copy of a previous cell'''
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split( X, y, test_size=0.2, random_state=42)

ValueError: Found input variables with inconsistent numbers of samples: [1994, 399]

> Your y and X did not have the same number of instances.

In [ ]:
Missing value treatment 2: replacing missing values with median of the respective feature

In [50]:
X = df_full.drop(['communityname','ViolentCrimesPerPop','highCrime'],axis=1)
X=X.apply(pd.to_numeric,errors='coerce')
X = X.fillna(X.median()).astype(int).astype(int)
y = df_full['highCrime']

In [51]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split( X, y, test_size=0.2, random_state=42)

>Now it ran this time, and I am not sure why it worked here, but it did not work previously.

In [52]:
from sklearn.model_selection import cross_val_score
accuracy =  cross_val_score(dtc, X_train, y_train, cv=10)
precision =  cross_val_score(dtc, X_train, y_train, cv=10 , scoring = 'precision')
recall =  cross_val_score(dtc, X_train, y_train, cv=10 , scoring = 'recall')
print ("Mean Accuracy = ", accuracy.mean())
print ("Mean Recall = ", recall.mean())
print ("Mean Precision = ", precision.mean())

Mean Accuracy =  0.668975549319
Mean Recall =  0.728159580664
Mean Precision =  0.744445031026


Decision tree could not be directly implemented with missing values. These missing values in the data had to be treated. There wer 2 approaches follwed:
1. Replacing the missing values with mean of the respective feature
2. Replacing the missing values with the median of the respective feature

After 10 cross fold validation, the accuracy for the 1st and the 2nd approach were 67.146 and 66.273 respectively. This is much lower than the accuracy that was gotten in the clean data, which was 76.94%. Hence we see that the results are worse with missing values and conclude that missing values have negative imoact on model.